### PR TITLE
Mississippi AGI

### PR DESCRIPTION
**Pull request by ZiqianCui**

### CODE REVIEW

File: changelog_entry.yaml [hunk 1]
```diff
@@ -0,0 +1,4 @@
+- bump: minor
+  changes:
+    added:
+    - MS agi.
```
Comment (PavelMakarchuk) [inline changelog_entry.yaml:4]:
```suggestion
    - Mississippi adjusted gross income.
```

File: policyengine_us/variables/gov/states/ms/tax/income/ms_agi.py [hunk 2]
```diff
@@ -0,0 +1,13 @@
+from policyengine_us.model_api import *
+
+
+class ca_agi(Variable):
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/ms/tax/income/ms_agi.py:4]:
```suggestion
class ms_agi(Variable):
```

File: policyengine_us/variables/gov/states/ms/tax/income/ms_agi_subtractions.py [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+class ms_agi_subtractions(Variable):
+    value_type = float
+    entity = TaxUnit
+    label = "MS AGI subtractions from federal AGI"
+    unit = USD
+    definition_period = YEAR
+    reference = "https://www.dor.ms.gov/sites/default/files/Forms/Individual/80100221.pdf#page=13"
+    defined_for = StateCode.MS
+
+    def formula(tax_unit, period, parameters):
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/ms/tax/income/ms_agi_subtractions.py:13]: need a subtractions .yaml folder

File: policyengine_us/variables/gov/states/ms/tax/income/ms_agi_subtractions.py [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+        agi = tax_unit("adjusted_gross_income", period)
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/ms/tax/income/ms_agi_subtractions.py:14]:
```suggestion
```

File: policyengine_us/variables/gov/states/ms/tax/income/ms_agi_subtractions.py [hunk 5]
```diff
@@ -0,0 +1,16 @@
+from policyengine_us.model_api import *
+
+
+class ms_agi_subtractions(Variable):
+    value_type = float
+    entity = TaxUnit
+    label = "MS AGI subtractions from federal AGI"
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/ms/tax/income/ms_agi_subtractions.py:7]:
```suggestion
    label = "Mississippi adjusted gross income subtractions"
```

File: policyengine_us/variables/gov/states/ms/tax/income/ms_agi_additions.py [hunk 6]
```diff
@@ -0,0 +1,11 @@
+from policyengine_us.model_api import *
+
+
+class ms_agi_additions(Variable):
+    value_type = float
+    entity = TaxUnit
+    label = "MS AGI additions to federal AGI"
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/ms/tax/income/ms_agi_additions.py:7]:
```suggestion
    label = "Mississippi adjusted gross income additions"
```

File: policyengine_us/variables/gov/states/ms/tax/income/ms_agi.py [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+    value_type = float
+    entity = TaxUnit
+    label = "Mississippi AGI"
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/ms/tax/income/ms_agi.py:7]:
```suggestion
    label = "Mississippi adjusted gross income"
```

File: policyengine_us/parameters/gov/states/ms/tax/income/subtractions/subtractions.yaml [hunk 8]
```diff
@@ -0,0 +1,10 @@
+description: Mississippi counts these sources as subtractions.
+values:
+  2021-01-01:
+    - ms_self_employment_subtraction
+
+metadata:
+  unit: list
+  reference: 
+    - title: Mississippi Resident INCOME TAX Instruction
+      href: https://www.dor.ms.gov/sites/default/files/Forms/Individual/80100221.pdf#page=12
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/ms/tax/income/subtractions/subtractions.yaml:10]: need a legal code

File: policyengine_us/variables/gov/states/ms/tax/income/ms_self_employment_subtraction.py
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/ms/tax/income/ms_self_employment_subtraction.py:1]: need a unit test for this variable

File: policyengine_us/variables/gov/states/ms/tax/income/ms_self_employment_subtraction.py [hunk 9]
```diff
@@ -0,0 +1,18 @@
+from policyengine_us.model_api import *
+
+
+class ms_self_employment_subtraction(Variable):
+    value_type = float
+    entity = TaxUnit
+    label = "Mississippi adjustemnts to federal adjusted gross income"
+    unit = USD
+    definition_period = YEAR
+    reference = "https://www.dor.ms.gov/sites/default/files/Forms/Individual/80100221.pdf#page=13"
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/ms/tax/income/ms_self_employment_subtraction.py:10]: need to add legal code

File: policyengine_us/tests/policy/baseline/gov/states/ms/tax/income/ms_agi_subtractions.yaml [hunk 10]
```diff
@@ -0,0 +1,7 @@
+- name: Mississippi agi
+  period: 2022
+  input:
+    state_code: MS
+    self_employment_tax: 20_000
+  output:
+    ms_agi_subtractions: 10_000
```
Comment (PavelMakarchuk) [inline policyengine_us/tests/policy/baseline/gov/states/ms/tax/income/ms_agi_subtractions.yaml:7]:
```suggestion
    ms_self_employment_subtraction: 10_000
```

File: policyengine_us/tests/policy/baseline/gov/states/ms/tax/income/ms_agi_subtractions.yaml
Comment (PavelMakarchuk) [inline policyengine_us/tests/policy/baseline/gov/states/ms/tax/income/ms_agi_subtractions.yaml:1]: add a test with no self-employment tax

File: policyengine_us/tests/policy/baseline/gov/states/ms/tax/income/ms_agi_subtractions.yaml [hunk 11]
```diff
@@ -0,0 +1,7 @@
+- name: Mississippi agi
```
Comment (PavelMakarchuk) [inline policyengine_us/tests/policy/baseline/gov/states/ms/tax/income/ms_agi_subtractions.yaml:1]:
```suggestion
- name: Calculation```

File: policyengine_us/parameters/gov/states/ms/tax/income/subtractions/self_employment/rate.yaml [hunk 12]
```diff
@@ -0,0 +1,10 @@
+description: Mississippi self employment tax deduction rate.
+values:
+  2022-10-01: 0.5
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/ms/tax/income/subtractions/self_employment/rate.yaml:3]: need a 2021 value

File: policyengine_us/parameters/gov/states/ms/tax/income/subtractions/self_employment/rate.yaml [hunk 13]
```diff
@@ -0,0 +1,10 @@
...
+metadata:
+  period: year
+  unit: /1
+  label: "Mississippi self employment tax deduction rate"
+  reference:
+  - title: Mississippi Resident INCOME TAX Instruction
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/ms/tax/income/subtractions/self_employment/rate.yaml:9]: need a legal code and 2022 tax form

File: policyengine_us/parameters/gov/states/ms/tax/income/income_sources.yaml
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/ms/tax/income/income_sources.yaml:1]: image

Comment (ZiqianCui) [reply policyengine_us/parameters/gov/states/ms/tax/income/income_sources.yaml:1]: still need line 41: rent, royalities and 48: other income to be added

File: policyengine_us/parameters/gov/states/ms/tax/income/income_sources.yaml [hunk 14]
```diff
@@ -0,0 +1,21 @@
+description: Mississippi counts these sources as income sources.
+values:
+  2021-01-01:
+    - e00200 # Wages, salaries, tips, etc.
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/ms/tax/income/income_sources.yaml:4]:
```suggestion
    - employment_income # Wages, salaries, tips, etc.
```

File: policyengine_us/parameters/gov/states/ms/tax/income/income_sources.yaml [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+    - qualified_dividend_income
+    - alimony_income
+    - taxable_pension_income
+    - unemployment_compensation
+metadata:
+  unit: list
+  reference:
+   - title: Mississippi Resident INCOME TAX Instruction
+      href: https://www.dor.ms.gov/sites/default/files/Forms/Individual/80100221.pdf#page=12
+  - title: 2020 Mississippi Code Title 27 - Taxation and Finance Chapter 7 - Income Tax and Withholding Article 1 - Income Tax
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/ms/tax/income/income_sources.yaml:19]: specific part of law

File: policyengine_us/parameters/gov/states/ms/tax/income/subtractions/self_employment/rate.yaml [hunk 16]
```diff
@@ -0,0 +1,12 @@
+description: Mississippi self employment tax deduction rate.
+values:
+  2021-10-01: 0.5
+metadata:
+  period: year
+  unit: /1
+  label: "Mississippi self employment tax deduction rate"
+  reference:
+  - title: Mississippi Resident INCOME TAX Instruction
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/ms/tax/income/subtractions/self_employment/rate.yaml:9]: add year and 2021 reference

File: policyengine_us/parameters/gov/states/ms/tax/income/subtractions/self_employment/rate.yaml [hunk 17]
```diff
@@ -0,0 +1,12 @@
+description: Mississippi self employment tax deduction rate.
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/ms/tax/income/subtractions/self_employment/rate.yaml:1]:
```suggestion
description: Mississippi multiplies the self employment taxes by this rate, under the self employment tax deduction.
```

File: policyengine_us/parameters/gov/states/ms/tax/income/subtractions/self_employment/rate.yaml [hunk 18]
```diff
@@ -0,0 +1,12 @@
...
+values:
+  2021-10-01: 0.5
+metadata:
+  period: year
+  unit: /1
+  label: "Mississippi self employment tax deduction rate"
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/ms/tax/income/subtractions/self_employment/rate.yaml:7]:
```suggestion
  label: Mississippi self employment tax deduction rate
```

File: policyengine_us/tests/policy/baseline/gov/states/ms/tax/income/ms_agi_subtractions.yaml [hunk 19]
```diff
@@ -0,0 +1,15 @@
+- name: Calculation of self-employment
+  period: 2021
+  input:
+    state_code: MS
+    self_employment_tax: 20_000
+  output:
+    ms_self_employment_subtraction: 10_000
+
+- name: Calculation of no self-employment
```
Comment (PavelMakarchuk) [inline policyengine_us/tests/policy/baseline/gov/states/ms/tax/income/ms_agi_subtractions.yaml:9]:
```suggestion
- name: Calculation without self-employment taxes
```

File: policyengine_us/tests/policy/baseline/gov/states/ms/tax/income/ms_agi_subtractions.yaml [hunk 20]
```diff
@@ -0,0 +1,15 @@
+- name: Calculation of self-employment
```
Comment (PavelMakarchuk) [inline policyengine_us/tests/policy/baseline/gov/states/ms/tax/income/ms_agi_subtractions.yaml:1]:
```suggestion
- name: Calculation with self-employment taxes
```

File: policyengine_us/variables/gov/states/ms/tax/income/ms_self_employment_subtraction.py [hunk 21]
```diff
@@ -0,0 +1,21 @@
+from policyengine_us.model_api import *
+
+
+class ms_self_employment_subtraction(Variable):
+    value_type = float
+    entity = TaxUnit
+    label = "Mississippi adjustemnts to federal adjusted gross income"
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/ms/tax/income/ms_self_employment_subtraction.py:7]:
```suggestion
    label = "Mississippi self employment tax deduction"
```

File: policyengine_us/variables/gov/states/ms/tax/income/ms_agi_subtractions.py [hunk 22]
```diff
@@ -0,0 +1,13 @@
+from policyengine_us.model_api import *
+
+
+class ms_agi_subtractions(Variable):
+    value_type = float
+    entity = TaxUnit
+    label = "Mississippi adjustemnts to federal adjusted gross income"
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/ms/tax/income/ms_agi_subtractions.py:7]:
```suggestion
    label = "Mississippi adjustments to federal adjusted gross income"
```

File: policyengine_us/parameters/gov/states/ms/tax/income/income_sources.yaml [hunk 23]
```diff
@@ -0,0 +1,23 @@
+description: Mississippi counts these sources as income sources.
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/ms/tax/income/income_sources.yaml:1]:
```suggestion
description: Mississippi accounts for these income sources.
```

File: policyengine_us/parameters/gov/states/ms/tax/income/income_sources.yaml [hunk 24]
```diff
@@ -0,0 +1,23 @@
...
+
+metadata:
+  unit: list
+  reference: 
+    - title: Mississippi Resident INCOME TAX Instruction
+      href: https://www.dor.ms.gov/sites/default/files/Forms/Individual/80100221.pdf#page=12 # Line 39 - 49
+    - title: Mississippi Income Tax and Withholding Article 1
+      href: https://law.justia.com/codes/mississippi/2020/title-27/chapter-7/article-1/section-27-7-15/
+    - title: Mississippi Resident Individual Income Tax Return 2022
+      href: https://www.dor.ms.gov/sites/default/files/Forms/Individual/80105228.pdf # Line 38 - 48
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/ms/tax/income/income_sources.yaml:23]: add page

File: policyengine_us/parameters/gov/states/ms/tax/income/subtractions/self_employment/rate.yaml [hunk 25]
```diff
@@ -0,0 +1,17 @@
...
+  period: year
+  unit: /1
+  label: Mississippi self employment tax deduction rate
+  reference:
+  - title: Mississippi Resident INCOME TAX Instruction
+    href: https://www.dor.ms.gov/sites/default/files/Forms/Individual/80100221.pdf#page=13
+  - title: 2020 Mississippi Code Title 27 Income Tax - self employment tax rate
+    href: https://law.justia.com/codes/mississippi/2020/title-27/chapter-7/article-1/section-27-7-18/ # 6(c)
+  - title: Mississippi Resident Individual Income Tax Return 2022
+    href: https://www.dor.ms.gov/sites/default/files/Forms/Individual/80105228.pdf  # Line 61
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/ms/tax/income/subtractions/self_employment/rate.yaml:16]: add page

File: policyengine_us/parameters/gov/states/ms/tax/income/subtractions/subtractions.yaml [hunk 26]
```diff
@@ -0,0 +1,14 @@
+description: Mississippi counts these sources as subtractions.
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/ms/tax/income/subtractions/subtractions.yaml:1]:
```suggestion
description: Mississippi counts these subtractions from the federal adjusted gross income.
```

File: policyengine_us/parameters/gov/states/ms/tax/income/subtractions/subtractions.yaml [hunk 27]
```diff
@@ -0,0 +1,14 @@
...
+values:
+  2021-01-01:
+    - ms_self_employment_subtraction
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/ms/tax/income/subtractions/subtractions.yaml:4]: comment out the subtractions that we have no included

File: policyengine_us/parameters/gov/states/ms/tax/income/subtractions/subtractions.yaml [hunk 28]
```diff
@@ -0,0 +1,14 @@
...
+
+metadata:
+  unit: list
+  reference: 
+    - title: Mississippi Resident INCOME TAX Instruction
+      href: https://www.dor.ms.gov/sites/default/files/Forms/Individual/80100221.pdf#page=12
+    - title: 2020 Mississippi Code Title 27 - Income Tax and Withholding Article 1
+      href: https://law.justia.com/codes/mississippi/2020/title-27/chapter-7/article-1/section-27-7-18/ # 1 - 7
+    - title: Mississippi Resident Individual Income Tax Return 2022
+      href: https://www.dor.ms.gov/sites/default/files/Forms/Individual/80105228.pdf  # Line 50 - 64
```
Comment (PavelMakarchuk) [inline policyengine_us/parameters/gov/states/ms/tax/income/subtractions/subtractions.yaml:14]: add page

File: policyengine_us/variables/gov/states/ms/tax/income/ms_agi_subtractions.py [hunk 29]
```diff
@@ -0,0 +1,16 @@
...
+
+
+class ms_agi_subtractions(Variable):
+    value_type = float
+    entity = TaxUnit
+    label = "Mississippi adjustments to federal adjusted gross income"
+    unit = USD
+    definition_period = YEAR
+    reference = (
+        "https://www.dor.ms.gov/sites/default/files/Forms/Individual/80100221.pdf#page=13",
```
Comment (PavelMakarchuk) [inline policyengine_us/variables/gov/states/ms/tax/income/ms_agi_subtractions.py:11]: I think if Mississippi calls them adjustments - we should rename everything to adjustments from subtractions

**Review by PavelMakarchuk — APPROVED** [review]

File: policyengine_us/variables/gov/states/ms/tax/income/ms_taxable_income.py [hunk 30]
```diff
@@ -7,4 +7,11 @@ class ms_taxable_income(Variable):
...
     unit = USD
     definition_period = YEAR
+    reference = (
+        "https://www.dor.ms.gov/sites/default/files/Forms/Individual/80100221.pdf#page=13",
+        "https://www.dor.ms.gov/sites/default/files/Forms/Individual/80105228.pdf",  # Line 38 - 49,
+    )
     defined_for = StateCode.MS
+
+    adds = ["ms_agi"]
+    subtracts = ["ms_deductions", "ms_total_exemptions"]
```
Comment (MaxGhenis) [inline policyengine_us/variables/gov/states/ms/tax/income/ms_taxable_income.py:17]: can this be negative? usually we do a max_(0, ...)

File: policyengine_us/variables/gov/states/ms/tax/income/adjustments/ms_self_employment_adjustment.py [hunk 31]
```diff
@@ -0,0 +1,21 @@
...
+    unit = USD
+    definition_period = YEAR
+    reference = [
+        "https://www.dor.ms.gov/sites/default/files/Forms/Individual/80100221.pdf#page=13",  # Line 61
+        "https://law.justia.com/codes/mississippi/2020/title-27/chapter-7/article-1/section-27-7-18/",  # 6(c)
+    ]
+    defined_for = StateCode.MS
+
+    def formula(tax_unit, period, parameters):
+        self_employment = add(tax_unit, period, ["self_employment_tax"])
```
Comment (MaxGhenis) [inline policyengine_us/variables/gov/states/ms/tax/income/adjustments/ms_self_employment_adjustment.py:17]: To distinguish from se income
```suggestion
        self_employment_tax = add(tax_unit, period, ["self_employment_tax"])
```

File: policyengine_us/parameters/gov/states/ms/tax/income/adjustments/adjustments.yaml [hunk 32]
```diff
@@ -0,0 +1,17 @@
+description: Mississippi counts these adjustments from the federal adjusted gross income.
```
Comment (MaxGhenis) [inline policyengine_us/parameters/gov/states/ms/tax/income/adjustments/adjustments.yaml:1]:
```suggestion
description: Mississippi subtracts these adjustments from federal adjusted gross income when computing state adjusted gross income.
```

File: policyengine_us/parameters/gov/states/ms/tax/income/adjustments/adjustments.yaml [hunk 33]
```diff
@@ -0,0 +1,17 @@
...
+values:
+  2020-01-01:
+    - ira_contributions # Line 50
+    - ms_self_employment_adjustment # Line 61
+    - ms_national_guard_or_reserve_pay_adjustment # Line 55
+    - alimony_expense # Line 53
+
+metadata:
+  unit: list
```
Comment (MaxGhenis) [inline policyengine_us/parameters/gov/states/ms/tax/income/adjustments/adjustments.yaml:10]:
```suggestion
  label: Mississippi AGI adjustments
  unit: list
```

File: policyengine_us/parameters/gov/states/ms/tax/income/adjustments/self_employment/rate.yaml [hunk 34]
```diff
@@ -0,0 +1,14 @@
+description: Mississippi deducts the following rate of self-employment taxes.
```
Comment (MaxGhenis) [inline policyengine_us/parameters/gov/states/ms/tax/income/adjustments/self_employment/rate.yaml:1]:
```suggestion
description: Mississippi subtracts this fraction of self-employment taxes from adjusted gross income.
```

File: policyengine_us/parameters/gov/states/ms/tax/income/income_sources.yaml [hunk 35]
```diff
@@ -0,0 +1,24 @@
...
+    - rent
+    - farm_income
+    - interest_income
+    - qualified_dividend_income
+    - alimony_income
+    - taxable_pension_income
+    - unemployment_compensation
+
+metadata:
+  unit: list
```
Comment (MaxGhenis) [inline policyengine_us/parameters/gov/states/ms/tax/income/income_sources.yaml:17]:
```suggestion
  label: Mississippi AGI income sources
  unit: list
```

File: policyengine_us/parameters/gov/states/ms/tax/income/income_sources.yaml [hunk 36]
```diff
@@ -0,0 +1,24 @@
+description: Mississippi accounts for these income sources.
```
Comment (MaxGhenis) [inline policyengine_us/parameters/gov/states/ms/tax/income/income_sources.yaml:1]:
```suggestion
description: Mississippi counts these income sources toward adjusted gross income.
```

**Review by MaxGhenis — CHANGES_REQUESTED** [review]

File: policyengine_us/variables/gov/states/ms/tax/income/ms_taxable_income.py (hunk 30 above)
Comment (PavelMakarchuk) [reply policyengine_us/variables/gov/states/ms/tax/income/ms_taxable_income.py]: yes, this should be negative

**Review by MaxGhenis — APPROVED** [review]

**Pull request merged**